src/main/scala/is/hail/linalg/BlockMatrix.scala

```diff
@@ -148,6 +149,149 @@ object BlockMatrix {
     }
   }
 
+  // rectangles checked in Python
+  def exportRectangles(
+    hc: HailContext,
+    input: String,
+    output: String,
+    flattenedRectangles: Array[Long],
+    delimiter: String,
+    nPartitions: Int): Unit = {
+    require(flattenedRectangles.length % 4 == 0)
+
+    val hadoop = hc.sc.hadoopConfiguration
+    if (!hadoop.exists(input + "/_SUCCESS"))
+      fatal("Write failed: no success indicator found")
+
+    val BlockMatrixMetadata(blockSize, nRows, nCols, maybeFiltered, partFiles) = BlockMatrix.readMetadata(hc, input)
+    val gp = GridPartitioner(blockSize, nRows, nCols, maybeFiltered)
+
+    val rectangles = flattenedRectangles.grouped(4).toArray    
+    val dRect = digitsNeeded(rectangles.length)
+
+    val sHadoopBc = hc.sc.broadcast(new SerializableHadoopConfiguration(hadoop))
+    val partFilesBc = hc.sc.broadcast(partFiles)
+
+    val rdd = hc.sc.parallelize(rectangles.zipWithIndex, numSlices = nPartitions).map { case (r, index) =>    
+      val partFiles = partFilesBc.value
+      val data = new Array[Double](blockSize)
+      val sb = new StringBuilder(blockSize << 2)
+      val paddedIndex = StringUtils.leftPad(index.toString, dRect, "0")
+      val outputFile = output + "/rect-" + paddedIndex + "_" + r.mkString("-")
+      val osw = new OutputStreamWriter(sHadoopBc.value.value.unsafeWriter(outputFile))
+
+      val startRow = r(0)
+      val stopRow = r(1)
+      val startCol = r(2)
+      val stopCol = r(3)
+      
+      val nonEmpty = startRow < stopRow && startCol < stopCol
+      
+      if (nonEmpty) {
+        val startRowOffset = gp.indexBlockOffset(startRow)
+  
+        val startBlockCol = gp.indexBlockIndex(startCol)
+        val startColOffset = gp.indexBlockOffset(startCol)
+  
+        val stopBlockCol = gp.indexBlockIndex(stopCol - 1) + 1
+        val stopColOffset = gp.indexBlockOffset(stopCol - 1) + 1
+  
+        val startColByteOffset = startColOffset << 3
+        val stopColByteDeficit = (gp.blockColNCols(stopBlockCol - 1) - stopColOffset) << 3
+  
+        val inPerBlockCol = new Array[InputBuffer](stopBlockCol - startBlockCol)
+  
+        var i = startRow
+        while (i < stopRow) {
+          if (i == startRow || gp.indexBlockOffset(i) == 0) {
+            val blockRow = gp.indexBlockIndex(i)
+            val nRowsInBlock = gp.blockRowNRows(blockRow)
+  
+            var blockCol = startBlockCol
+            while (blockCol < stopBlockCol) {
+              val pi = gp.coordinatesPart(blockRow, blockCol)
+              if (pi < 0)
+                fatal(s"block ($blockRow, $blockCol) missing for rectangle $index " +
+                  s"with bounds ${ r.mkString("[", ", ", "]") }")
+  
+              val is = sHadoopBc.value.value.unsafeReader(input + "/parts/" + partFiles(pi))
```

Comment:
The closing should be in a try / finally, in case something in the middle errors out

Reply:
it also feels like this method could be tighter by iterating through the blocks and only creating one active input stream at a time.

Reply:
Do you see a way to do this without opening and closing an input stream for every row of every block?

Reply:
did you have something like this in mind?
```
          try {
         ...
          } catch {
            case e: Exception =>
              inPerBlockCol.foreach(_.close())
              throw e
          } finally {
            if (i % blockSize == 0 || i == stopRow)
              inPerBlockCol.foreach(_.close())
          }
```

Reply:
actually, that may result in trying to close twice.